source/controllers/include/controllers/IController.hpp

```diff
@@ -0,0 +1,100 @@
+#pragma once
+
+#include <list>
+#include <map>
+#include <memory>
+
+#include "controllers/exceptions/NoRobotModelException.hpp"
+
+#include "robot_model/Model.hpp"
+#include "state_representation/parameters/ParameterMap.hpp"
+#include "state_representation/space/Jacobian.hpp"
+#include "state_representation/space/joint/JointPositions.hpp"
+#include "state_representation/space/joint/JointState.hpp"
+
+/**
+ * @namespace controllers
+ * @brief Systems to determine a command output from measured and desired inputs
+ */
+namespace controllers {
+/**
+ * @class IController
+ * @brief Abstract class to define a controller in a desired state type, such as joint or Cartesian spaces
+ * @tparam S the state type of the controller
+ */
+template<class S>
+class IController : public state_representation::ParameterMap {
+public:
+  /**
+   * @brief Empty constructor
+   */
+  IController() = default;
+
+  /**
+   * @brief Empty destructor
+   */
+  virtual ~IController() = default;
+
+  /**
+   * @brief Compute the command output based on the commanded state and a feedback state
+   * To be redefined based on the actual controller implementation.
+   * @param command_state the input state to the controller
+   * @param feedback_state the current state of the system given as feedback
+   * @return the output command at the input state
+   */
+  [[nodiscard]] virtual S compute_command(const S& command_state, const S& feedback_state) = 0;
+
+  /**
+   * @brief Compute the command output in joint space
+   * @param command_state
+   * @param feedback_state
+   * @param jacobian the Jacobian matrix relating Cartesian forces to joint space
+   * @return
+   */
+  [[nodiscard]] state_representation::JointState compute_command(
+      const S& command_state, const S& feedback_state, const state_representation::Jacobian& jacobian
+  );
+
+  /**
+   * @brief Compute the command output in joint space from command and feedback states in task space
+   * @details This method requires the controller to have an associated robot model to compute the Jacobian.
+   * @param command_state
+   * @param feedback_state
+   * @param joint_positions The current joint positions, which are required for computing the Jacobian
+   * @param frame_name
+   * @return
+   */
+  [[nodiscard]] state_representation::JointState compute_command(
+      const S& command_state, const S& feedback_state, const state_representation::JointPositions& joint_positions,
+      const std::string& frame_name = ""
+  );
+
+  /**
+   * @brief Get the robot model associated with the controller.
+   * @return The robot model
+   */
+  [[nodiscard]] const robot_model::Model& get_robot_model();
+
+  /**
+   * @brief Set the robot model associated with the controller.
+   * @param robot_model The robot model
+   */
+  void set_robot_model(const robot_model::Model& robot_model);
```

Comment:
I read your comment in the README regarding changing the model at runtime and I am now wondering if we should actually allow that. You could remove the setter and just accept the getter which would make it more immutable. We could even potentially declare it as a const attribute and set it from the constructor, although maybe some of the robot model functions are actually mutating it.

Reply:
The problem is, passing the robot model through the constructor would require all derived classes to reimplement such a constructor. I also don't expect run-time changes to the robot model to really happen all that much, and it has to be done explicitly by the user - it's not a shared reference or anything but rather a concrete object (copied from the const reference). 

One useful case might be where we want to update the mass associated with a robot model for crba dynamics, because it picked up a payload or whatever, but all the joints and reference frames stay the same. This would be a case where the user can set a new robot model in between computations. Even if the reference frame names of the robot model are changed, it won't break anything as long as the states passed to `compute_command` also have the same changed reference frame. 

The main "dangerous" casesI mentioned would be changing the number of joints in an initialised joint controller. This would lead to dimensionality mismatch _if_ the robot model were actually used for anything, for example to check joint limits.

In most cases, setting robot model post construction is an occasional convenience that can, in some cases, be used in a dumb way and cause an exception. The user cannot always be prevented from doing things wrong, just like you would get an exception if you tried to multiply incompatible states together. Here I think the note warning is sufficient and I wouldn't change the logic.

Reply:
I see the point. I am fine like this then. Thanks for clarifying.